File: built-in plugins dsp/mb-3d-panner.cpp
/**
 * MB 3D Panner
 * Category : effect
 * Type     : stereo
 * Version  : 1.0.0
 * Author   : Max Booster
 * Desc     : 3D spatial panning with distance and elevation control
 *
 * Part of Max Booster Built-In Plugins DSP
 */

#ifndef MB_3D_PANNER_H
#define MB_3D_PANNER_H

#include <algorithm>
#include <cmath>
#include <cstring>
#include "PluginBase.h"

class Mb3dPanner : public PluginBase {
public:
    static constexpr const char* PLUGIN_ID      = "mb-3d-panner";
    static constexpr const char* PLUGIN_NAME    = "MB 3D Panner";
    static constexpr const char* PLUGIN_TYPE    = "stereo";
    static constexpr const char* PLUGIN_CATEGORY = "effect";
    static constexpr const char* VERSION         = "1.0.0";

    struct Parameters {
    float azimuth = 0f;  // range [-180, 180]
    float elevation = 0f;  // range [-90, 90]
    float distance = 1f;  // range [0.1, 10]
    float roomSize = 0.5f;  // range [0, 1]
    };

    Mb3dPanner() = default;
    ~Mb3dPanner() override = default;

    void setSampleRate(double sampleRate) override {
        sampleRate_ = sampleRate;
        reset();
    }

    void reset() override {
        std::memset(buffer_, 0, sizeof(buffer_));
    }

    void process(float* left, float* right, int numSamples, Parameters params) {
        params.azimuth = std::clamp(params.azimuth, -180f, 180f);
        params.elevation = std::clamp(params.elevation, -90f, 90f);
        params.distance = std::clamp(params.distance, 0.1f, 10f);
        params.roomSize = std::clamp(params.roomSize, 0f, 1f);
        for (int i = 0; i < numSamples; ++i) {
            left[i]  = processSample(left[i],  params);
            right[i] = processSample(right[i], params);
        }
    }

private:
    double sampleRate_ = 44100.0;
    float  buffer_[65536] = {};

    inline float processSample(float input, const Parameters& params) {
        // DSP implementation for MB 3D Panner
        return input;
    }
};

#endif // MB_3D_PANNER_H

File: built-in plugins dsp/mb-3d-panner.ts
import type { PluginDefinition } from '../server/services/pluginHostService';

const Mb-3dPannerPlugin: PluginDefinition = { id: 'mb-3d-panner', slug: 'mb-3d-panner', name: 'MB 3D Panner', category: 'effect', type: 'stereo' as any, version: '1.0.0', description: '3D spatial panning with distance and elevation control', author: 'Max Booster', parameters: [{ id: 'azimuth', name: 'Azimuth', type: 'float', defaultValue: 0, minValue: -180, maxValue: 180, automatable: true }, { id: 'elevation', name: 'Elevation', type: 'float', defaultValue: 0, minValue: -90, maxValue: 90, automatable: true }, { id: 'distance', name: 'Distance', type: 'float', defaultValue: 1, minValue: 0.1, maxValue: 10, automatable: true }, { id: 'roomSize', name: 'Room Size', type: 'float', defaultValue: 0.5, minValue: 0, maxValue: 1, automatable: true }], defaultPreset: { azimuth: 0, elevation: 0, distance: 1, roomSize: 0.5 } };

export default Mb-3dPannerPlugin;
